substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1062 @@
+			epoch_tag ^= 1;
+			slot_idx -= epoch_len;
+			if tickets_meta.unsorted_tickets_count != 0 {
+				Self::sort_tickets(u32::MAX, epoch_tag, &mut tickets_meta);
+				TicketsMeta::<T>::set(tickets_meta);
+			}
+		} else if slot_idx >= 2 * epoch_len {
+			return None
+		}
+
+		let ticket_idx = get_ticket_idx(slot_idx);
+		if ticket_idx < tickets_meta.tickets_count[epoch_tag as usize] {
+			TicketsIds::<T>::get((epoch_tag, ticket_idx))
+		} else {
+			None
+		}
+	}
+
+	/// Returns ticket id and data associated to the given `slot`.
+	///
+	/// Refer to the `slot_ticket_id` documentation for the slot-ticket association
+	/// criteria.
+	pub fn slot_ticket(slot: Slot) -> Option<(TicketId, TicketBody)> {
+		Self::slot_ticket_id(slot).and_then(|id| TicketsData::<T>::get(id).map(|body| (id, body)))
+	}
+
+	// Lexicographically sort the tickets which belong to the next epoch.
+	//
+	// Tickets are fetched from at most `max_segments` segments.
+	//
+	// The resulting sorted vector is optionally truncated to contain at most `MaxTickets`
+	// entries. If all the unsorted segments are consumed then the sorted vector is
+	// saved as the next epoch tickets, else it is saved to be used by next calls to
+	// this function.
+	pub(crate) fn sort_tickets(max_segments: u32, epoch_tag: u8, metadata: &mut TicketsMetadata) {
+		let mut unsorted_segments_count =
+			metadata.unsorted_tickets_count.div_ceil(SEGMENT_MAX_SIZE);
+		let max_segments = max_segments.min(unsorted_segments_count);
+		let max_tickets = MaxTicketsFor::<T>::get() as usize;
+
+		// Fetch the sorted candidates (if any).
+		let mut sorted_candidates = SortedCandidates::<T>::take().into_inner();
+
+		let mut require_sort = max_segments != 0;
+
+		// There is an upper bound to check only if we already sorted the max number
+		// of allowed tickets.
+		let mut upper_bound = *sorted_candidates.get(max_tickets - 1).unwrap_or(&TicketId::MAX);
+
+		// Consume at most `max_segments` segments.
+		// During the process remove every stale ticket from `TicketsData` storage.
+		for _ in 0..max_segments {
+			unsorted_segments_count -= 1;
+			let segment = UnsortedSegments::<T>::take(unsorted_segments_count);
+			metadata.unsorted_tickets_count -= segment.len() as u32;
+
+			// Push only ids with a value less than the current `upper_bound`.
+			// As ticket ids follow a uniform random distribution we expect to
+			// drop more or less half of the segment tickets here.
+			segment.into_iter().for_each(|id| {
+				if id < upper_bound {
+					sorted_candidates.push(id);
+				} else {
+					TicketsData::<T>::remove(id);
+				}
+			});
+
+			if sorted_candidates.len() > max_tickets {
+				// Sort, truncate good tickets, cleanup storage.
+				require_sort = false;
+				sorted_candidates.sort_unstable();
+				sorted_candidates[max_tickets..]
+					.iter()
+					.for_each(|id| TicketsData::<T>::remove(id));
+				sorted_candidates.truncate(max_tickets);
+				upper_bound = sorted_candidates[max_tickets - 1];
+			}
```

Comment:
I tried to benchmark using the [`sort_segments`](https://github.com/paritytech/polkadot-sdk/blob/ffefc226b4810efab226fdc5892492b7dd1cc912/substrate/frame/sassafras/src/benchmarking.rs#L163) benchmark I wrote (constructs and sorts up to 100 segments of 128 tickets)

Some summary results (`max_tickets` is set to 200, `x` is segments count):
- as is (i.e. always sort and trunc): 178 + x·163 µs
- sort when above 1024: 180 + x·163 µs
- sort when above 10000: 181.8 + x·166 µs

Doesn't look there is much difference (from the benches is even "slower" 1µs 🤣), just guessing... but may be because of the upper bound that is updated to discard even more tickets during the process

Reply:
What I want to do maybe is to implement the `UnsortedSegment` as a queue. I.e. start sorting from the first segment instead of the last.

If we perform incremental sorting while we continue receiving new tickets then this may be smarter as it always start from "full" segments (the last is typically not full)

Reply:
> just guessing... but may be because of the upper bound that is updated to discard even more tickets during the process

Could someone deliberately spam a bunch of tickets which are always less than the upper bound? Can you try with such tickets and see if there's a difference there?

> tried to benchmark using the [`sort_segments`](https://github.com/paritytech/polkadot-sdk/blob/ffefc226b4810efab226fdc5892492b7dd1cc912/substrate/frame/sassafras/src/benchmarking.rs#L163) benchmark I wrote

FYI, when you want to compare performance using benchmarks it's better to not look at the weights and instead get the benchmarking machinery to emit a `.json` with the raw times and just use those. The weights are a result of linear interpolation, so there are cases where they can completely lie to you (especially if e.g. the execution time is not actually linear).

Reply:
No spamming is not possible. Ticket Id value is out of control of the user. Is the result of a vrf.

Reply:
For the raw values you are talking about. I also looked into it

Reply:
> FYI, when you ...

I drop the results here as a reference

- *as is* https://pastebin.com/1tpDUxwA
- sort threshold 1000 https://pastebin.com/U7sr1CMi
- sort threshold 10000 https://pastebin.com/jfYMcPQZ